Unable to compile on Python 3.11 alpha 5

When running `python3.11 -m pip wheel hpy --use-pep517` the output is as follows:

```console
Collecting hpy
  Downloading hpy-0.0.3.tar.gz (172 kB)
     |████████████████████████████████| 172 kB 9.5 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: hpy
  Building wheel for hpy (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/domdf/.local/bin/python3.11 /home/domdf/python311/lib/python3.11/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /tmp/tmp4xeykvcx
       cwd: /tmp/pip-wheel-9n1000e1/hpy_bf00928872e2496a9d116bd1dd260d92
  Complete output (37 lines):
  fatal: not a git repository (or any of the parent directories): .git
  running bdist_wheel
  running build
  running build_py
  running egg_info
  writing hpy.egg-info/PKG-INFO
  writing dependency_links to hpy.egg-info/dependency_links.txt
  writing entry points to hpy.egg-info/entry_points.txt
  writing requirements to hpy.egg-info/requires.txt
  writing top-level names to hpy.egg-info/top_level.txt
  listing git files failed - pretending there aren't any
  reading manifest file 'hpy.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  running build_ext
  building 'hpy.universal' extension
  hpy/debug/src/debug_ctx_cpython.c: In function ‘debug_ctx_CallRealFunctionFromTrampoline’:
  hpy/debug/src/debug_ctx_cpython.c:162:20: error: dereferencing pointer to incomplete type ‘cpy_Py_buffer’ {aka ‘struct bufferinfo’}
    162 |             a->view->obj = NULL;
        |                    ^~
  hpy/debug/src/debug_ctx_cpython.c:165:36: warning: passing argument 3 of ‘_buffer_h2py’ from incompatible pointer type [-Wincompatible-pointer-types]
    165 |         _buffer_h2py(dctx, &hbuf, a->view);
        |                                   ~^~~~~~
        |                                    |
        |                                    cpy_Py_buffer * {aka struct bufferinfo *}
  hpy/debug/src/debug_ctx_cpython.c:41:78: note: expected ‘Py_buffer *’ {aka ‘struct <anonymous> *’} but argument is of type ‘cpy_Py_buffer *’ {aka ‘struct bufferinfo *’}
     41 | static void _buffer_h2py(HPyContext *dctx, const HPy_buffer *src, Py_buffer *dest)
        |                                                                   ~~~~~~~~~~~^~~~
  hpy/debug/src/debug_ctx_cpython.c:173:29: warning: passing argument 2 of ‘_buffer_py2h’ from incompatible pointer type [-Wincompatible-pointer-types]
    173 |         _buffer_py2h(dctx, a->view, &hbuf);
        |                            ~^~~~~~
        |                             |
        |                             cpy_Py_buffer * {aka struct bufferinfo *}
  hpy/debug/src/debug_ctx_cpython.c:56:61: note: expected ‘const Py_buffer *’ {aka ‘const struct <anonymous> *’} but argument is of type ‘cpy_Py_buffer *’ {aka ‘struct bufferinfo *’}
     56 | static void _buffer_py2h(HPyContext *dctx, const Py_buffer *src, HPy_buffer *dest)
        |                                            ~~~~~~~~~~~~~~~~~^~~
  error: command '/usr/bin/gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for hpy
Failed to build hpy
ERROR: Failed to build one or more wheels

```

The output is more or less the same when building from git, with some of the line numbers changing.

-----

It looks like `bufferinfo` was removed in https://github.com/python/cpython/pull/29991, which made it into alpha 5.

It's used here:

https://github.com/hpyproject/hpy/blob/dd702f55ce391df43cbf5dab1ce7d4b1be23110b/hpy/devel/include/hpy/cpy_types.h#L12

Just tried to build with Python 3.11.0 and got this:
```
hpy/devel/src/runtime/ctx_type.c:657:58: error: declaration of 'struct bufferinfo' will not be visible outside of this function [-Werror,-Wvisibility]
typedef int (*HPyGetBufferProc)(struct _object *, struct bufferinfo *, int);
                                                         ^
hpy/devel/src/runtime/ctx_type.c:658:63: error: declaration of 'struct bufferinfo' will not be visible outside of this function [-Werror,-Wvisibility]
typedef void (*HPyReleaseBufferProc)(struct _object *, struct bufferinfo *);
                                                              ^
hpy/devel/src/runtime/ctx_type.c:677:48: error: incompatible function pointer types assigning to 'getbufferproc' (aka 'int (*)(struct _object *, Py_buffer *, int)') from 'HPyGetBufferProc' (aka 'int (*)(struct _object *, struct bufferinfo *, int)') [-Werror,-Wincompatible-function-pointer-types]
                    buffer_procs->bf_getbuffer = (HPyGetBufferProc)src->slot.cpy_trampoline;
                                               ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hpy/devel/src/runtime/ctx_type.c:687:52: error: incompatible function pointer types assigning to 'releasebufferproc' (aka 'void (*)(struct _object *, Py_buffer *)') from 'HPyReleaseBufferProc' (aka 'void (*)(struct _object *, struct bufferinfo *)') [-Werror,-Wincompatible-function-pointer-types]
                    buffer_procs->bf_releasebuffer = (HPyReleaseBufferProc)src->slot.cpy_trampoline;
                                                   ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 errors generated
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -22,6 +22,8 @@ jobs:
           - os: ubuntu-latest
             python-version: '3.10'
             compiler: 'g++'
+          - os: ubuntu-latest
+            python-version: '3.11'
 
     steps:
       - uses: actions/checkout@v2
@@ -59,7 +61,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        python-version: ['3.10', '3.9', '3.8', '3.7']
+        python-version: ['3.11', '3.10', '3.9', '3.8', '3.7']
 
     steps:
       - uses: actions/checkout@v2
@@ -98,6 +100,8 @@ jobs:
             python-version: '3.8'
           - os: ubuntu-latest
             python-version: '3.9'
+          - os: ubuntu-latest
+            python-version: '3.11'
 
     steps:
       - uses: actions/checkout@v2

--- a/hpy/devel/src/runtime/ctx_type.c
+++ b/hpy/devel/src/runtime/ctx_type.c
@@ -652,10 +652,16 @@ create_slot_defs(HPyType_Spec *hpyspec, HPy_ssize_t base_member_offset,
     return result;
 }
 
-// XXX: This is a hack to work-around the missing Py_bf_getbuffer and
-// Py_bf_releasebuffer before 3.9. We shouldn't use it on 3.9+.
-typedef int (*HPyGetBufferProc)(struct _object *, struct bufferinfo *, int);
-typedef void (*HPyReleaseBufferProc)(struct _object *, struct bufferinfo *);
+/* Python 3.8 and older is the missing Py_bf_getbuffer and Py_bf_releasebuffer
+   so we need to define those functions here. Since Python 3.9, we can just use
+   the function pointer types from Python. */
+#if PY_VERSION_HEX < 0x03009000
+typedef int (*cpy_getbufferproc)(cpy_PyObject *, cpy_Py_buffer *, int);
+typedef void (*cpy_releasebufferproc)(cpy_PyObject *, cpy_Py_buffer *);
+#else
+typedef getbufferproc cpy_getbufferproc;
+typedef releasebufferproc cpy_releasebufferproc;
+#endif
 static PyBufferProcs*
 create_buffer_procs(HPyType_Spec *hpyspec)
 {
@@ -674,7 +680,7 @@ create_buffer_procs(HPyType_Spec *hpyspec)
                             return NULL;
                         }
                     }
-                    buffer_procs->bf_getbuffer = (HPyGetBufferProc)src->slot.cpy_trampoline;
+                    buffer_procs->bf_getbuffer = (cpy_getbufferproc)src->slot.cpy_trampoline;
                     break;
                 case HPy_bf_releasebuffer:
                     if (buffer_procs == NULL) {
@@ -684,7 +690,7 @@ create_buffer_procs(HPyType_Spec *hpyspec)
                             return NULL;
                         }
                     }
-                    buffer_procs->bf_releasebuffer = (HPyReleaseBufferProc)src->slot.cpy_trampoline;
+                    buffer_procs->bf_releasebuffer = (cpy_releasebufferproc)src->slot.cpy_trampoline;
                     break;
                 default:
                     break;

--- a/hpy/devel/src/runtime/ctx_type.c
+++ b/hpy/devel/src/runtime/ctx_type.c
@@ -158,7 +158,7 @@ static HPyType_Extra_t *_HPyType_Extra_Alloc(const char *name, HPyType_BuiltinSh
     memcpy(result->name, name, name_size);
     result->shape = shape;
     result->magic = HPy_TYPE_MAGIC;
-    /* XXX the returned struct is never freed */
+    /* XXX On Python 3.10 and older, the returned struct is never freed */
     return result;
 }
 
@@ -1135,6 +1135,18 @@ ctx_Type_FromSpec(HPyContext *ctx, HPyType_Spec *hpyspec,
     if (result == NULL) {
         return HPy_NULL;
     }
+
+#if PY_VERSION_HEX >= 0x030B0000
+    /* Since Python 3.11, the name will be copied into a fresh buffer such that
+       the type objects owns this buffer. Hence we need to patch the pointer
+       again. The nice thing is that Python will then free out HPyType_Extra
+       for us. The internal field '_ht_tpname' is only used for freeing. So, we
+       can set it to 'extra'. */
+    PyHeapTypeObject *ht = (PyHeapTypeObject *) result;
+    PyMem_Free(ht->_ht_tpname);
+    ht->_ht_tpname = (char *) extra;
+    ht->ht_type.tp_name = (const char *)extra->name;
+#endif
 
 #if PY_VERSION_HEX < 0x03080000
     /*

